langchain/callbacks/mlflow_callback.py

```diff
@@ -0,0 +1,692 @@
+        resp.update(self.metrics)
+
+        self.mlflg.metrics(self.metrics, step=self.metrics["step"])
+
+        for generations in response.generations:
+            for idx, generation in enumerate(generations):
+                generation_resp = deepcopy(resp)
+                generation_resp.update(flatten_dict(generation.dict()))
+                generation_resp.update(
+                    analyze_text(
+                        generation.text,
+                        nlp=self.nlp,
+                    )
+                )
+                self.mlflg.metrics(
+                    generation_resp.pop("text_complexity_metrics"),
+                    step=self.metrics["step"],
+                )
+                self.records["on_llm_end_records"].append(generation_resp)
+                self.records["action_records"].append(generation_resp)
+                self.mlflg.jsonf(resp, f"llm_end_{llm_ends}_generation_{idx}")
+                dependency_tree = generation_resp["dependency_tree"]
+                entities = generation_resp["entities"]
+                self.mlflg.html(dependency_tree, "dep-" + hash_string(generation.text))
+                self.mlflg.html(entities, "ent-" + hash_string(generation.text))
+
+    def on_llm_error(
+        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
+    ) -> None:
+        """Run when LLM errors."""
+        self.metrics["step"] += 1
+        self.metrics["errors"] += 1
+
+    def on_chain_start(
+        self, serialized: Dict[str, Any], inputs: Dict[str, Any], **kwargs: Any
+    ) -> None:
+        """Run when chain starts running."""
+        self.metrics["step"] += 1
+        self.metrics["chain_starts"] += 1
+        self.metrics["starts"] += 1
+
+        chain_starts = self.metrics["chain_starts"]
+
+        resp = {}
+        resp.update({"action": "on_chain_start"})
+        resp.update(flatten_dict(serialized))
+        resp.update(self.metrics)
+
+        self.mlflg.metrics(self.metrics, step=self.metrics["step"])
+
+        chain_input = inputs["input"]
+
+        if isinstance(chain_input, str):
+            input_resp = deepcopy(resp)
+            input_resp["input"] = chain_input
+            self.records["on_chain_start_records"].append(input_resp)
+            self.records["action_records"].append(input_resp)
+            self.mlflg.jsonf(input_resp, f"chain_start_{chain_starts}")
+        elif isinstance(chain_input, list):
+            for idx, inp in enumerate(chain_input):
+                input_resp = deepcopy(resp)
+                input_resp.update(inp)
+                self.records["on_chain_start_records"].append(input_resp)
+                self.records["action_records"].append(input_resp)
+                self.mlflg.jsonf(input_resp, f"chain_start_{chain_starts}_input_{idx}")
+        else:
+            raise ValueError("Unexpected data format provided!")
+
+    def on_chain_end(self, outputs: Dict[str, Any], **kwargs: Any) -> None:

```

Comment:
Shall we log the model using the mlflow langchain flavor? https://github.com/mlflow/mlflow/blob/master/mlflow/langchain/__init__.py

Reply:
yes @liangz1

Reply:
@liangz1 added support for mlflow langchain flavor.